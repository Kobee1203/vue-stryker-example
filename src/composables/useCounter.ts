import { ref } from 'vue'

const MIN_VALUE = 0
const MAX_VALUE = 10

export const useCounter = (initialCounter: number = MIN_VALUE) => {
  let initialValue = initialCounter

  // AStryker disable next-line EqualityOperator: The <= mutant results in an equivalent mutant
  if (initialCounter < 0) {
    console.debug('Initial counter is %s, setting it to 0', initialCounter)
    initialValue = 0
  }
  // AStryker disable next-line EqualityOperator: The >= mutant results in an equivalent mutant
  if (initialCounter > 10) {
    console.debug('Initial counter is %s, setting it to 10', initialCounter)
    initialValue = 10
  }

  const count = ref(initialValue)

  function increment() {
    if (count.value < MAX_VALUE) {
      count.value++
    }
  }

  function decrement() {
    if (count.value > MIN_VALUE) {
      count.value--
    }
  }

  return {
    count,
    increment,
    decrement,
  }
}
