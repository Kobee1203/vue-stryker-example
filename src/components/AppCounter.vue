<script setup lang="ts">
import { useCounter } from '@/composables/useCounter.ts'
import { computed } from 'vue'

const props = withDefaults(defineProps<{initialCounter: number}>(), {
  initialCounter: 0,
})

const { count, decrement, increment } = useCounter(props.initialCounter)

const countClass = computed(() => (count.value > 5 ? 'text-red-500' : 'text-green-500'))
</script>

<template>
  <button class="button-decrement" @click="decrement">-</button>
  <span class="count-value" :class="countClass">{{ count }}</span>
  <button class="button-increment" @click="increment">+</button>
</template>

<style scoped>
.count-value {
  margin: 0 1rem;
}

.text-red-500 {
  color: #ef4444;
}

.text-green-500 {
  color: #22c55e;
}
</style>
