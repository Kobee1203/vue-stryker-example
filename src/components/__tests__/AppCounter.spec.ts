import { describe, expect, it, vi } from 'vitest'

import { mount } from '@vue/test-utils'
import AppCounter from '../AppCounter.vue'
import { useCounter } from '@/composables/useCounter.ts'
import { ref } from 'vue'

const incrementMock = vi.fn()
const decrementMock = vi.fn()

const useCounterMock = {
  count: ref(0),
  increment: incrementMock,
  decrement: decrementMock,
}

vi.mock('@/composables/useCounter.ts', () => ({
  useCounter: vi.fn(),
}))

describe('AppCounter', () => {
  it.each([0, 1, 2, 3, 4, 5])('renders properly', (initialCounter: number) => {
    vi.mocked(useCounter).mockReturnValue({ ...useCounterMock, count: ref(0) })

    const wrapper = mount(AppCounter, { props: { initialCounter }})
    expect(wrapper.html()).matchSnapshot()
    expect(useCounter).toHaveBeenCalledWith(initialCounter)
  })

  it.each([0, 1, 2, 3, 4, 5])(
    'renders the text in green when the count value is less or equal than 5',
    (countValue: number) => {
      vi.mocked(useCounter).mockReturnValue({ ...useCounterMock, count: ref(countValue) })

      const wrapper = mount(AppCounter, { props: { initialCounter: 0 }})
      //const button = wrapper.find('.button-increment')
      //button.trigger('click')
      expect(wrapper.html()).matchSnapshot()
    },
  )

  it.each([6, 7, 8, 9, 10])(
    'renders the text in red when the count value is greater than 5',
    (countValue: number) => {
      vi.mocked(useCounter).mockReturnValue({ ...useCounterMock, count: ref(countValue) })

      const wrapper = mount(AppCounter, { props: { initialCounter: 0 }})
      //const button = wrapper.find('.button-increment')
      //button.trigger('click')
      expect(wrapper.html()).matchSnapshot()
    },
  )
})
