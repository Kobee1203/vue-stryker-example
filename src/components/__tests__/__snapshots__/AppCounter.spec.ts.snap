// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppCounter > renders properly 1`] = `
"<button data-v-84e281ac="" class="button-decrement">-</button>
<span data-v-84e281ac="" class="count-value text-green-500">0</span>
<button data-v-84e281ac="" class="button-increment">+</button>"
`;

exports[`AppCounter > renders properly 2`] = `
"<button data-v-84e281ac="" class="button-decrement">-</button>
<span data-v-84e281ac="" class="count-value text-green-500">0</span>
<button data-v-84e281ac="" class="button-increment">+</button>"
`;

exports[`AppCounter > renders properly 3`] = `
"<button data-v-84e281ac="" class="button-decrement">-</button>
<span data-v-84e281ac="" class="count-value text-green-500">0</span>
<button data-v-84e281ac="" class="button-increment">+</button>"
`;

exports[`AppCounter > renders properly 4`] = `
"<button data-v-84e281ac="" class="button-decrement">-</button>
<span data-v-84e281ac="" class="count-value text-green-500">0</span>
<button data-v-84e281ac="" class="button-increment">+</button>"
`;

exports[`AppCounter > renders properly 5`] = `
"<button data-v-84e281ac="" class="button-decrement">-</button>
<span data-v-84e281ac="" class="count-value text-green-500">0</span>
<button data-v-84e281ac="" class="button-increment">+</button>"
`;

exports[`AppCounter > renders properly 6`] = `
"<button data-v-84e281ac="" class="button-decrement">-</button>
<span data-v-84e281ac="" class="count-value text-green-500">0</span>
<button data-v-84e281ac="" class="button-increment">+</button>"
`;

exports[`AppCounter > renders the text in green when the count value is less or equal than 5 1`] = `
"<button data-v-84e281ac="" class="button-decrement">-</button>
<span data-v-84e281ac="" class="count-value text-green-500">0</span>
<button data-v-84e281ac="" class="button-increment">+</button>"
`;

exports[`AppCounter > renders the text in green when the count value is less or equal than 5 2`] = `
"<button data-v-84e281ac="" class="button-decrement">-</button>
<span data-v-84e281ac="" class="count-value text-green-500">1</span>
<button data-v-84e281ac="" class="button-increment">+</button>"
`;

exports[`AppCounter > renders the text in green when the count value is less or equal than 5 3`] = `
"<button data-v-84e281ac="" class="button-decrement">-</button>
<span data-v-84e281ac="" class="count-value text-green-500">2</span>
<button data-v-84e281ac="" class="button-increment">+</button>"
`;

exports[`AppCounter > renders the text in green when the count value is less or equal than 5 4`] = `
"<button data-v-84e281ac="" class="button-decrement">-</button>
<span data-v-84e281ac="" class="count-value text-green-500">3</span>
<button data-v-84e281ac="" class="button-increment">+</button>"
`;

exports[`AppCounter > renders the text in green when the count value is less or equal than 5 5`] = `
"<button data-v-84e281ac="" class="button-decrement">-</button>
<span data-v-84e281ac="" class="count-value text-green-500">4</span>
<button data-v-84e281ac="" class="button-increment">+</button>"
`;

exports[`AppCounter > renders the text in green when the count value is less or equal than 5 6`] = `
"<button data-v-84e281ac="" class="button-decrement">-</button>
<span data-v-84e281ac="" class="count-value text-green-500">5</span>
<button data-v-84e281ac="" class="button-increment">+</button>"
`;

exports[`AppCounter > renders the text in red when the count value is greater than 5 1`] = `
"<button data-v-84e281ac="" class="button-decrement">-</button>
<span data-v-84e281ac="" class="count-value text-red-500">6</span>
<button data-v-84e281ac="" class="button-increment">+</button>"
`;

exports[`AppCounter > renders the text in red when the count value is greater than 5 2`] = `
"<button data-v-84e281ac="" class="button-decrement">-</button>
<span data-v-84e281ac="" class="count-value text-red-500">7</span>
<button data-v-84e281ac="" class="button-increment">+</button>"
`;

exports[`AppCounter > renders the text in red when the count value is greater than 5 3`] = `
"<button data-v-84e281ac="" class="button-decrement">-</button>
<span data-v-84e281ac="" class="count-value text-red-500">8</span>
<button data-v-84e281ac="" class="button-increment">+</button>"
`;

exports[`AppCounter > renders the text in red when the count value is greater than 5 4`] = `
"<button data-v-84e281ac="" class="button-decrement">-</button>
<span data-v-84e281ac="" class="count-value text-red-500">9</span>
<button data-v-84e281ac="" class="button-increment">+</button>"
`;

exports[`AppCounter > renders the text in red when the count value is greater than 5 5`] = `
"<button data-v-84e281ac="" class="button-decrement">-</button>
<span data-v-84e281ac="" class="count-value text-red-500">10</span>
<button data-v-84e281ac="" class="button-increment">+</button>"
`;
